### PR TITLE
On-demand table backup introduced.

### PR DESCRIPTION
**Pull request by radkomateusz**
It performs on-demand backup for single non partitioned table or single partition of partitioned table, without taking care about shouldBackup function results.

it could be used to enforce making backup of given table or partition.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -143,6 +143,17 @@ It's worth to underline that:
 * Backups for partitions are scheduled randomly within the range of time specified in [config.yaml](./config/config.yaml),
 * It is possible to check the progress via [Task Queues](https://console.cloud.google.com/appengine/taskqueues).
 
+### On-demand table backup
+BBQ provides option for performing on-demand backup for specified table or partition.
```
Comment (marcin-kolda) [inline README.md]: I'm wondering if we want to document it, as we haven't documented Advanced section at all, i.e. restore from list. We could put some of the docs to the UI tooltips.

File: templates/on_demand_table_backup.html [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      document.getElementById("response").innerHTML = "Request timeout: " + e;
+    };
+
+    xhttp.send();
+    return false;
+  }
+</script>
+<h2>On-demand table backup</h2>
+<p class="lead">
+  BBQ will perform on-demand backup of given source table. It performs backup regardless of result returned from shouldBackup predicate.
```
Comment (marcin-kolda) [inline templates/on_demand_table_backup.html]: I would describe what is ignored. End user shouldn't need to check what's in predicate.

File: templates/on_demand_table_backup.html [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      <input type="number" class="form-control" id="sourcePartitionId" name="partitionId"
+             placeholder="partition id">
+      <small id="sourcePartitionIdHelp" class="form-text text-muted">Source partition id for partitioned table
+      </small>
+    </div>
+  </div>
+  </div>
+  <div class="form-group row">
+    <div class="col-sm-4">
+      <button type="submit" class="btn btn-primary">Perform backup</button>
```
Comment (marcin-kolda) [inline templates/on_demand_table_backup.html]: Maybe "Schedule backup" would be better?

File: README.md (hunk 1 above)
Comment (radkomateusz) [reply README.md]: I agree that UI should inform user almost about everything, but on the other hand, if feature is already described in README, why should we remove it?

File: templates/on_demand_table_backup.html (hunk 2 above)
Comment (radkomateusz) [reply templates/on_demand_table_backup.html]: done

File: templates/on_demand_table_backup.html (hunk 3 above)
Comment (radkomateusz) [reply templates/on_demand_table_backup.html]: done

File: src/backup/default_should_backup_predicate.py [hunk 4]
```diff
@@ -1,40 +1,35 @@
...
+        if self._table_has_up_to_date_backup(big_query_table_metadata, table_entity):
             logging.info('Backup is up to date')
             return False
+
         return True
 
-    # pylint: disable=R0201
-    def __should_backup(self, table_entity):
+    def _table_has_up_to_date_backup(self, big_query_table_metadata, table_entity):
         # TODO: change name of this class or split this method into two
```
Comment (marcin-kolda) [inline src/backup/default_should_backup_predicate.py]: Is this TODO still needed?

File: src/backup/on_demand/on_demand_should_backup_predicate.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        if not self._is_possible_to_copy_table(big_query_table_metadata):
+            return False
+
+        if big_query_table_metadata.is_empty():
+            logging.info('This table is empty')
+
+        logging.info(
+            "Performing on-demand backup for %s:%s.%s$%s. "
+            "It is performed without checking if table aready has up to date backup",
+            table_entity.project_id, table_entity.dataset_id,
```
Comment (marcin-kolda) [inline src/backup/on_demand/on_demand_should_backup_predicate.py]: You can print it as `big_query_metadata.table_reference()`

File: src/backup/on_demand/on_demand_should_backup_predicate.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    AbstractShouldBackupPredicate
+
+
+class OnDemandShouldBackupPredicate(AbstractShouldBackupPredicate):
+
+    def test(self, big_query_table_metadata, table_entity):
+        if not self._is_possible_to_copy_table(big_query_table_metadata):
+            return False
+
+        if big_query_table_metadata.is_empty():
```
Comment (marcin-kolda) [inline src/backup/on_demand/on_demand_should_backup_predicate.py]: Can we move this log to abstract method?

File: src/backup/abstract_should_backup_predicate.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import logging
+from abc import abstractmethod
+
+
+class AbstractShouldBackupPredicate(object):
```
Comment (marcin-kolda) [inline src/backup/abstract_should_backup_predicate.py]: Can we remove Should from `AbstractShouldBackupPredicate` and `OnDemandShouldBackupPredicate`

File: templates/on_demand_table_backup.html [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    xhttp.send();
+    return false;
+  }
+</script>
+<h2>On-demand table backup</h2>
+<p class="lead">
+  BBQ will schedule on-demand backup of given source table. <br>
+  On-demand flow doesn't check if table already has up to date backup<br>
+
+  Note that successfully scheduled on-demand table backup could fail during executing copy job. You need to check results via looking in application logs.
```
Comment (marcin-kolda) [inline templates/on_demand_table_backup.html]: "via looking" -> "by looking"?

File: src/backup/abstract_should_backup_predicate.py (hunk 7 above)
Comment (radkomateusz) [reply src/backup/abstract_should_backup_predicate.py]: done

File: src/backup/on_demand/on_demand_should_backup_predicate.py (hunk 6 above)
Comment (radkomateusz) [reply src/backup/on_demand/on_demand_should_backup_predicate.py]: done

File: templates/on_demand_table_backup.html (hunk 8 above)
Comment (radkomateusz) [reply templates/on_demand_table_backup.html]: done

File: src/backup/default_should_backup_predicate.py (hunk 4 above)
Comment (radkomateusz) [reply src/backup/default_should_backup_predicate.py]: refactored. done

**Pull request merged**